libraries/AP_UAVCAN/AP_UAVCAN_Servers.cpp

```diff
@@ -0,0 +1,318 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Siddharth Bharat Purohit
+ */
+
+/****************************************************************************
+*
+*   Copyright (c) 2015 PX4 Development Team. All rights reserved.
+*      Author: Pavel Kirienko <pavel.kirienko@gmail.com>
+*              David Sidrane <david_s5@usa.net>
+*
+****************************************************************************/
+
+
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN
+
+#include "AP_UAVCAN_Servers.h"
+
+#ifdef HAS_UAVCAN_SERVERS
+
+#include <uavcan/protocol/dynamic_node_id_server/event.hpp>
+#include <uavcan/protocol/dynamic_node_id_server/storage_backend.hpp>
+#include <uavcan/protocol/dynamic_node_id_server/centralized.hpp>
+#include <uavcan/protocol/HardwareVersion.hpp>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include <AP_BoardConfig/AP_BoardConfig_CAN.h>
+
+#include <AP_Common/AP_Common.h>
+#include <DataFlash/DataFlash.h>
+
+#if HAL_OS_POSIX_IO
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <stdio.h>
+#include <time.h>
+#include <dirent.h>
+#if defined(__APPLE__) && defined(__MACH__)
+#include <sys/param.h>
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#endif
+#endif
+
+#if HAL_OS_FATFS_IO
+#include <stdio.h>
+#endif
+
+#ifndef UAVCAN_NODE_DB_PATH
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
+#define UAVCAN_NODE_DB_PATH "/fs/microsd/APM/UAVCAN"
+#else
+#define UAVCAN_NODE_DB_PATH "/APM/UAVCAN"
+#endif
+#endif
+
+#ifndef UAVCAN_LOG_FILE
+#define UAVCAN_LOG_FILE     UAVCAN_NODE_DB_PATH"/trace.log"
+#endif
+
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_uavcan(fmt, args...) do { hal.console->printf(fmt, ##args); } while (0)
+
+class AP_UAVCAN_CentralizedServer : public uavcan::dynamic_node_id_server::CentralizedServer
+{
+public:
+    AP_UAVCAN_CentralizedServer(uavcan::INode& node, uavcan::dynamic_node_id_server::IStorageBackend &storage_backend, uavcan::dynamic_node_id_server::IEventTracer &tracer) :
+        uavcan::dynamic_node_id_server::CentralizedServer(node, storage_backend, tracer) {}
+};
+
+class AP_UAVCAN_FileEventTracer : public uavcan::dynamic_node_id_server::IEventTracer
+{
+protected:
+    virtual void onEvent(uavcan::dynamic_node_id_server::TraceCode code, uavcan::int64_t argument)
+    {
+        struct log_UAVCAN_EVT pkt = {
+            LOG_PACKET_HEADER_INIT(LOG_UAVCAN_EVT_MSG),
+            time_us    : AP_HAL::micros64(),
+            code       : code,
+            argument   : argument
+        };
+        DataFlash_Class::instance()->WriteBlock(&pkt, sizeof(pkt));
+    }
```

Comment:
it doesn't, this just puts data into buffer. IO writing is done in its own thread.

Reply:
Yeah, but putting into the buffer has to be protected. I checked the code and at least on WriteBlock a Semaphore is used, so I think this should be OK.

Reply:
I think change this to use the newer Log_Write() logging method, which means we don't use a logging slot for boards not using CAN, and also makes the code a bit more readable.
Also note that logging is safe from any thread now (as are mavlink sends)

Reply:
@tridge Not using a logging slot for CAN just makes it easier to run out of slots and fail silently at logging during runtime when we are out of slots. We panic in SITL, but since most SITL runs don't enable CAN it's very unlikely to be caught, we need to not be banking on Log_Write everywhere, or we need to refactor of 8 bit slot ID's.

Reply:
I disagree, if we followed this approach we'll run out of slots much earlier
while we are stuck with 8 bit, we should dynamically allocated the IDs using Log_Write() for low-bandwidth logging like this

Reply:
I see Sid has changed it to Log_Write() ... thanks!